### ISSUE TITLE
`@classmethod` does not type check with decorated methods

### ISSUE THREAD

**Issue by moosbruggerj** [microsoft/pyright#3937]
**Describe the bug**
Using `@classmethod` on a, with a descriptor, decorated method reports an error, that classmethod needs a callable as argument.
It type checks when using the `@property` decorator instead of any arbitrary class decorator (#1423).

**To Reproduce**
See code snipped below.

**Expected behavior**
This code should type check.

**Screenshots or Code**

```python
from typing import Any, Callable

class descriptor:
    def __init__(self, func: Callable[..., Any]):
        self.func = func
    def __get__(self, obj, cls=None):
        def _wrapper(*args, **kwargs):
            if hasattr(self.func, '__get__'):
                _f = self.func.__get__(obj, cls)
            else:
                _f = self.func
            return _f(*args, **kwargs)
        return _wrapper
class Test:
    @classmethod  # <- error is reported here
    @descriptor
    def cls_method(cls):
        print(cls)

Test.cls_method()
```
**Error message**
```
Argument of type "descriptor" cannot be assigned to parameter "__f" of type
"(...) -> _R_co@classmethod" in function "__init__"
  Type "descriptor" cannot be assigned to type "(...) -> _R_co@classmethod"
(Pyright reportGeneralTypeIssues)

```

**VS Code extension or command-line**
Neovim
coc-pyright 1.1.270 with Pyright 1.1.270

**Additional context**
This snipped is reported as correct when using mypy (version: `mypy 0.961 (compiled: yes)`).

**Comment by erictraut** [comment]
`@classmethod` must be applied to a function (a callable object). It cannot be applied to a non-function object such as a descriptor.

If you look at the definition of `classmethod` in `builtins.pyi`, you'll see that its `__init__` method requires a parameter of type `Callable`.

You can fix the issue in your code by swapping the two decorators.

**Comment by moosbruggerj** [comment]
Thank you. I will change my code accordingly.

The current situation seems confusing though (at least to me).
The documentation of 3.10 specifically calls this out as valid python code, in both the `classmethod` and the `descriptor` docs and there is no mention of it being deprecated in 3.11.
I only learnt this through your comment on a similar issue, as I usually don't follow changes in the next python version before it is released.
Then there is the special case of the `property` descriptor that currently is accepted by pyright. Will this be removed as well going forward?
Lastly, mypy accepts the whole pattern.

I know both the docs and mypy are out of scope of this repository, but would it be possible/feasible to add an error message to pyright that explains the situation with the deprecation and such?
Or do you think this too much of a niche issue?
